src/base/components/usb_logger.rs

```diff
@@ -0,0 +1,103 @@
+// Based on https://github.com/DCRichards/arduino-vagrant-rust/blob/7bdd4a5eb15639a78cf7826a95863e21a98eb0bb/apps/plant/app/src/usb_logger.rs
+use {
+    arduino_nano33iot as bsp,
+    bsp::hal::clock::GenericClockController,
+    bsp::hal::usb::UsbBus,
+    bsp::pac::{interrupt, PM, USB},
+    bsp::usb_allocator,
+    core::fmt::{Error, Write},
+    core::result::Result,
+    cortex_m::interrupt::free,
+    cortex_m::peripheral::NVIC,
+    usb_device::bus::UsbBusAllocator,
+    usb_device::prelude::*,
+    usbd_serial::{SerialPort, USB_CLASS_CDC},
+};
+
+macro_rules! from {
+    ($stone:expr) => {
+        base::components::usb_logger::USBLogger::new(
+            $stone.usb,
+            &mut $stone.clock,
+            &mut $stone.pm,
+            $stone.pins.usb_dm.into(),
+            $stone.pins.usb_dp.into(),
+            &mut $stone.nvic,
+        )
+    };
+}
+pub(crate) use from;
+
+static mut USB_ALLOCATOR: Option<UsbBusAllocator<UsbBus>> = None;
+static mut USB_BUS: Option<UsbDevice<UsbBus>> = None;
+static mut USB_SERIAL: Option<SerialPort<UsbBus>> = None;
+
+/// Represents a logger which logs over USB Serial.
+pub struct USBLogger {}
+
+impl USBLogger {
+    /// Get the log instance
+    pub fn new(
+        usb: USB,
+        clocks: &mut GenericClockController,
+        pm: &mut PM,
+        dm: bsp::UsbDm,
+        dp: bsp::UsbDp,
+        nvic: &mut bsp::pac::NVIC,
+    ) -> Self {
+        let bus_allocator = unsafe {
+            USB_ALLOCATOR = Some(usb_allocator(usb, clocks, pm, dm, dp));
+            USB_ALLOCATOR.as_ref().unwrap()
+        };
+
+        unsafe {
+            USB_SERIAL = Some(SerialPort::new(bus_allocator));
+            USB_BUS = Some(
+                // vendor id, product id
+                UsbDeviceBuilder::new(bus_allocator, UsbVidPid(0x2341, 0x8057))
+                    .manufacturer("Arduino LLC")
+                    .product("Arduino NANO 33 IoT")
+                    .serial_number("4197FF4750535131332E3120FF0B1541")
+                    .device_class(USB_CLASS_CDC)
+                    .build(),
+            );
+            
+            nvic.set_priority(interrupt::USB, 1);
+            // Enable the interrupt in the NVIC.
+            NVIC::unmask(interrupt::USB);
+        }
+
+        USBLogger {}
+    }
+
+    /// Log writes a log entry
+    pub fn log(&self, s: &str) {
```

Comment:
Again, unless we can fix the `static mut` issue... this shouldn't need to take `&self` as it's unused.

Reply:
Let's rewrite all this file :)